app.server/server.ts

```diff
@@ -1,114 +1,134 @@
-
-  // get all the products out of the query result, ignoring the collection they're in
-  let products = _.flatMap(queryResult.collections, c => c.products)
-
-  let realWmsUrl = getRealWmsUrl(app.settings.env, req.header(`Host`) || '', req.protocol)
-  let result = getCapabilities(products, realWmsUrl)
-
-  res.set(`Content-Type`, `text/xml`)
-  res.send(result)
-})
-
-// return collections to the ui
-app.get(`/api/collections`, (req, res) => {
-  let result = getCollections()
-  res.json(result)
-})
-
-// return products to the ui
-app.get(`/api/products`, (req, res) => {
-  let q = parseQuerystring(req.query)
-  validateQuery(q)
-  let result = getProducts(q)
-  res.json(result)
-})
-
-// store the query and give me a key for it
-app.post(`/api/storedQueries`, async (req, res) => {
-  let query = req.body
-  validateQuery(query)
-  let storedQuery = await storedQueryRepository.store(query)
-  res.json({ key: storedQuery.id })
-})
-
-// single page app. any routes that are "pages" need to return the index.html
-// and allow the client-side router to show the correct page
-app.get(pages, (req, res) => {
-  console.log(req.url)
-  res.sendFile('index.html', { root: env.dir })
-})
-// serve our html pages
-// (note: webpack-dev-server serves the index.html pages from in-memory and doesn't reach this server)
-// for (let page of pages) {
-//   app.get(page.path, (req, res) => {
-//     res.sendFile(page.file, { root: env.dir })
-//   })
-// }
-
-// exercise error handling
-app.get(`/error`, (req, res) => {
-  throw 'You made an error!'
-})
-
-app.get(`/500`, (req, res) => {
-  res.sendFile('errors/500.html', { root: env.dir })
-})
-
-// serve static files from the specified directory
-app.use(express.static(env.dir))
-
-// no matches yet, return 404
-app.use((req, res) => {
-  res.status(404)
-  res.sendFile('errors/404.html', { root: env.dir })
-})
-
-// error, return 500
-if (!env.dev) {
-  app.use((err: any, req: any, res: any, next: any) => {
-    res.status(500)
-    res.sendFile('errors/500.html', { root: env.dir })
-  })
-}
-
-// start the express web server
-app.listen(env.port, () => {
-  console.log(`it's ` + new Date().toISOString())
-  console.log(`app.server is listening on: http://localhost:${env.port}`)
-  console.log(`node environment is ${env.name}`)
-})
-
-let errorHandler: express.ErrorRequestHandler = (err, req, res, next) => {
-  console.error(err.stack)
-  res.status(500).send('Something broke!')
-}
+let catalog = new Catalog();
+
+// TODO: Returns a finished app at the end of an async call to build a
+// list of OGC layers for the known collections. This shouldn't be here
+// but we need to marry the list of OGC services in the OGC collection
+// to the list of lidar collections, so need to know which collection
+// maps to which OGC layer at runtime, lazy loading didn't work well
+// for some reason so it lives here for the moment.
+exports.appPromise = catalog.getOGCServiceList()
```

Comment:
I don't understand what `exports.appPromise` is for ?

Reply:
its just a way of returning the finished app, the last command in the file is returned in this case appPromise which resolves to the app variable. This was just what was suggested in this pattern.

Reply:
Why do we need to export anything at all though? And what uses the value `appPromise`?

Reply:
Nothing does at present, its a way of getting the system to wait while something happens before returning the completed app and thus bringing the server online. The logic here being that if we change the startup logic in anyway this is how you would create the app, i.e. resolve the promise to get a finished app.

General idea came from https://stackoverflow.com/questions/41364072/running-an-async-function-before-express-js-start

Reply:
Would like to understand this better but let's adjourn for now.